bundles/org.openhab.binding.upnpcontrol/src/main/java/org/openhab/binding/upnpcontrol/internal/handler/UpnpServerHandler.java

```diff
@@ -0,0 +1,471 @@
+
+    private @NonNullByDefault({}) ChannelUID rendererChannelUID;
+    private @NonNullByDefault({}) ChannelUID currentTitleChannelUID;
+
+    private volatile int numberReturned;
+    private volatile int totalMatches;
+
+    private volatile UpnpEntry currentEntry = new UpnpEntry(DIRECTORY_ROOT, DIRECTORY_ROOT, DIRECTORY_ROOT,
+            "object.container");
+    private volatile List<UpnpEntry> entries = new ArrayList<>(); // current entry list in selection
+    private volatile Map<String, UpnpEntry> parentMap = new HashMap<>(); // store parents in hierarchy separately to be
+                                                                         // able to move up in directory structure
+
+    private List<String> source = new ArrayList<>();
+
+    private UpnpDynamicStateDescriptionProvider upnpStateDescriptionProvider;
+    private UpnpDynamicCommandDescriptionProvider upnpCommandDescriptionProvider;
+
+    public UpnpServerHandler(Thing thing, UpnpIOService upnpIOService,
+            ConcurrentMap<String, UpnpRendererHandler> upnpRenderers,
+            UpnpDynamicStateDescriptionProvider upnpStateDescriptionProvider,
+            UpnpDynamicCommandDescriptionProvider upnpCommandDescriptionProvider) {
+        super(thing, upnpIOService);
+        this.upnpRenderers = upnpRenderers;
+        this.upnpStateDescriptionProvider = upnpStateDescriptionProvider;
+        this.upnpCommandDescriptionProvider = upnpCommandDescriptionProvider;
+
+        // put root as highest level in parent map
+        parentMap.put(currentEntry.getId(), currentEntry);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+
+        logger.debug("Initializing handler for media server device {}", thing.getLabel());
+
+        rendererChannelUID = thing.getChannel(UPNPRENDERER).getUID();
+        currentTitleChannelUID = thing.getChannel(BROWSE).getUID();
+
+        if (config.udn != null) {
+            if (service.isRegistered(this)) {
+                initServer();
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Communication cannot be established with " + thing.getLabel());
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "No UDN configured for " + thing.getLabel());
+        }
+    }
+
+    private void initServer() {
+        rendererStateOptionList = new ArrayList<>();
+        upnpRenderers.forEach((key, value) -> {
+            StateOption stateOption = new StateOption(key, value.getThing().getLabel());
+            rendererStateOptionList.add(stateOption);
+        });
+        updateStateDescription(rendererChannelUID, rendererStateOptionList);
+
+        getProtocolInfo();
+
+        browse(currentEntry.getId(), "BrowseDirectChildren", "*", "0", "0", getConfig().get(SORT_CRITERIA).toString());
```

Comment:
The config variables should be stored in a variable in the `initialize()` method. You can define a class with the variables and use `getConfigAs`.